symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```diff
@@ -0,0 +1,98 @@
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..), oneM)
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), evalPolynomial, var)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.MerkleHash   (merkleHash, runHash)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize ::
+   (Arithmetic a, Ord (Rep i), Functor o, Binary (Rep i)) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize ac = let (newAcSystem, vs) = varsToReplace (acSystem ac, M.empty) in optimizeAc vs ac {acSystem = newAcSystem}
+
+varsToReplace ::
+  (Arithmetic a, Ord (Rep i)) =>
+  (Map ByteString (Constraint a i) , Map (SysVar i) a) -> (Map ByteString (Constraint a i) , Map (SysVar i) a)
+varsToReplace (s, l) = let newVars = M.fromList . M.elems $ mapMaybe toConstVar s in
+  if newVars == M.empty
+  then (s, l)
+  else varsToReplace (optimizeSystems newVars s, M.union newVars l)
+
+optimizeSystems :: (Arithmetic a, Ord (Rep i)) =>
+  Map (SysVar i) a -> Map ByteString (Constraint a i) -> Map ByteString (Constraint a i)
+optimizeSystems m as = M.filter (/= zero) $ evalPolynomial evalMonomial varF <$> as
+  where
+    varF p = maybe (var p) fromConstant (M.lookup p m)
+
+optimizeAc :: forall a p i o. (Arithmetic a, Ord (Rep i), Functor o, Binary (Rep i))
+  => Map (SysVar i) a-> ArithmeticCircuit a p i o -> ArithmeticCircuit a p i o
+optimizeAc vs ac =
+    ac {
+      acSystem = addInVarConstraints $ acSystem ac,
+      acRange =  MM.filter (/= S.empty) $ optRanges vs (acRange ac),
+      acWitness = (>>= optWitVar vs) <$> M.filterWithKey (\k _ -> notMember (NewVar k) vs) (acWitness ac),
+      acOutput = acOutput ac <&> \case
+        SysVar sV -> maybe (SysVar sV) ConstVar (M.lookup sV vs)
+        o -> o
+      }
+  where
+    addInVarConstraints :: Map ByteString (Poly a (SysVar i) Natural) -> Map ByteString (Poly a (SysVar i) Natural)
+    addInVarConstraints p = foldrWithKey (\k a as -> case k of
+      inVar@(InVar inV) -> M.insert (runHash @(Just (Order a)) $ merkleHash inV) (var inVar - fromConstant a) as
+      _                 -> as) p vs
+
+    optRanges :: Map (SysVar i) a -> MM.MonoidalMap a (S.Set (SysVar i)) -> MM.MonoidalMap a (S.Set (SysVar i))
+    optRanges m = MM.mapWithKey (\r s -> bool (error "range constraint less then value") (S.difference s $ keysSet m) (all (<= r) $ restrictKeys m s))
```

Comment:
I think `mapMaybeWithKey` might be a better fit here (helps to get rid of `filter` afterwards)